File: frontend/src/types/document.ts
import type { User } from "./user";

export interface Document {
  id: string;
  title: string;
  content?: string; // JSON string of Quill content
  creator: User;
  projectId: string;
  createdAt: string;
  updatedAt: string;
}

export interface DocumentListItem {
  id: string;
  title: string;
  creator: User;
  projectId: string;
  createdAt: string;
  updatedAt: string;
}

export interface DocumentCreationRequest {
  title: string;
  content?: string;
  projectId: string;
}

export interface DocumentUpdateRequest {
  title?: string;
  content?: string;
}

export interface DocumentResponse {
  message: string;
  result: Document;
}

// Backend trả về DocumentListResponse trực tiếp trong ApiResponse.result
export interface DocumentListResponse {
  id: string;
  title: string;
  creator: User;
  projectId: string;
  createdAt: string;
  updatedAt: string;
}

File: backend/src/main/java/com/example/backend/entity/Document.java
package com.example.backend.entity;

import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.FieldDefaults;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.JsonNode;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.annotations.Type;

import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Builder
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE)
@Table(name = "documents")
public class Document {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    UUID id;

    @Column(name = "title", nullable = false)
    private String title;

    @Column(name = "content", columnDefinition = "jsonb")
    @Type(com.vladmihalcea.hibernate.type.json.JsonType.class)
    private JsonNode content;

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "creator_id")
    User creator;

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "project_id")
    Project project;

    @CreationTimestamp
    @Column(name = "created_at")
    OffsetDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    OffsetDateTime updatedAt;
}
